Disentangle implicit-selection logic from specificity rules and make …

milessabin opened the pull request:
…implicit selection more consistent with normal overload resolution.

This is an attempt to make implicit selection rules align more closely with people's intuitions and the spec's aspirations that it be consistent with normal overload resolution rules. It also fixes https://github.com/scala/bug/issues/10526.

Candidate implicits are ordered via `Infer#isStrictlyMoreSpecific` which delegates to `Infer#isAsSpecific`. Prior to this PR the latter treats all methods with an implicit argument list as if they were nullary methods, leading to https://github.com/scala/bug/issues/10526 ... I believe that the motivation for this behaviour was to allow methods which are intended to appear to users to be nullary to have implicit arguments without affecting their overload resolution behaviour wrt same-named methods which are _actually_ nullary. Whilst this does the right thing in some circumstances, it leads to oddities such as https://github.com/scala/bug/issues/10526 and also means that the specificity test used to order and select implicits is very crude, in effect _only_ considering the result type of an implicit method, not its implicit arguments.

This PR addresses these issues by making the "treat implicit methods as nullary" logic explicitly flagged and enabling it only at the relevant call site in `Inferencer#inferExprAlternative`. Normal overloading rules are unchanged, and the implicit selection logic is extracted out to `ImplicitSearch#Improves`. There the relative ordering of a pair of candidate implicits is computed as follows,
1. The candidates are compared by their result type, the most specific by normal overload rules being selected (this is consistent with the behaviour prior to this PR).
2. If (1) is a tie then we compare the lengths of the implicit argument lists of the candidates (if any). The candidate with the longest argument list wins (this matches users intuitions that methods with more implicit arguments are "more demanding" hence more specific).
3. If (2) is a tie then we have a pair of candidates with implicit argument lists of equal length and which are both equally specific wrt their result type/the expected type. We now apply the full normal overloading resolution rules and choose the most specific.

This change makes combining implicits from multiple implicit scopes a great deal more straightforward, since we can now rely on more than just the result type of the implicit definition to guide implicit selection. After this PR the following works as expected,
```scala
class Show[T](val i: Int)
object Show {
  def apply[T](implicit st: Show[T]): Int = st.i

  implicit val showInt: Show[Int] = new Show[Int](0)
  implicit def fallback[T]: Show[T] = new Show[T](1)
}

class Generic
object Generic {
  implicit val gen: Generic = new Generic
  implicit def showGen[T](implicit gen: Generic): Show[T] = new Show[T](2)
}

object Test extends App {
  assert(Show[Int] == 0)
  assert(Show[String] == 1)
  assert(Show[Generic] == 2) // showGen beats fallback due to longer argument list
}
```
Thanks to the use of normal overload resolution rules we also have the ability to use a "phantom" implicit parameter with arguments ordered by subtyping to explicitly prioritize implicit definitions, something for which language extensions of one form or another have occasionally been proposed,
```scala
class Low
object Low {
  implicit val low: Low = new Low
}
class Medium extends Low
object Medium {
  implicit val medium: Medium = new Medium
}
class High extends Medium
object High {
  implicit val high: High = new High
}

class Foo[T](val i: Int)
object Foo {
  def apply[T](implicit fooT: Foo[T]): Int = fooT.i

  implicit def foo[T](implicit priority: Low): Foo[T] = new Foo[T](0)
  implicit def foobar[T](implicit priority: Low): Foo[Bar[T]] = new Foo[Bar[T]](1)
  implicit def foobarbaz(implicit priority: Low): Foo[Bar[Baz]] = new Foo[Bar[Baz]](2)
}
class Bar[T]
object Bar {
  implicit def foobar[T](implicit priority: Medium): Foo[Bar[T]] = new Foo[Bar[T]](3)
  implicit def foobarbaz(implicit priority: Medium): Foo[Bar[Baz]] = new Foo[Bar[Baz]](4)
}
class Baz
object Baz {
  implicit def baz(implicit priority: High): Foo[Bar[Baz]] = new Foo[Bar[Baz]](5)
}

object Test extends App {
  assert(Foo[Int] == 0)
  assert(Foo[Bar[Int]] == 3)
  assert(Foo[Bar[Baz]] == 5)
}
```
This PR changes the outcome of only one partest (`test/files/neg/t7289_status_quo.scala`) which is intended to exactly capture the current failure behaviour in a specific implicit resolution scenario ... I believe the new (successful) behaviour is an improvement on the old.

With that exception all (par)tests pass. One shapeless test fails to compile after this change but is trivially fixable. I'll schedule a community build when this goes green ...

milessabin commented:
As expected the [community build](https://scala-ci.typesafe.com/job/scala-2.13.x-integrate-community-build/729/) fails due to a single shapeless test failure. Everything not downstream of shapeless builds successfully however, which is very promising.

lrytz commented on test/files/neg/t7289_status_quo.scala:
```diff
@@ -12,12 +12,15 @@ object Test1 {
-  implicitly[Ext[_ <: List[List[Int]]]]         // fails - ambiguous
+  implicitly[Ext[_ <: List[List[Int]]]]         // compiles due to f and m having equally specific result types due
+                                                // to scala/bug#10545 but f now being selected over m because it has
+                                                // the longer implicit argument list. When the fix for 10545 is
```
shouldn't the bug reference be 10526 here?

milessabin replied:
No, that's the right bug reference ... scala/bug#10526 and scala/bug#10545 intersect here.

milessabin replied:
See https://github.com/scala/scala/pull/6122.

lrytz replied:
Ah got it, sorry for the noise.

lrytz commented on src/compiler/scala/tools/nsc/typechecker/Implicits.scala:
```diff
@@ -401,18 +401,58 @@ trait Implicits {
+
+      def compute = {
+        // 1. Do we win on result type specificity?
+        val result0 = isStrictlyMoreSpecific(restpe1, restpe2, info1.sym, info2.sym, false)
```
style nit: use a named argument when passing a literal for a boolean param (also in the other invocations of `isStrictlyMoreSpecific`)

milessabin replied:
I'll fix.

lrytz reviewed (commented):
The fix for scala/bug#10526 looks good to me. About the new ranking rules, should that come with a spec update? I see that the current spec is already vague, but there are new rules now that are quite different from what we have in overloading resolution. It looks like we should bring that proposal to people's attention first - what is dotty doing?

milessabin commented:
@lrytz I've pointed @smarter and @odersky at this ... some interest, but no very definite response as yet. I think this is consistent with @odersky's intentions to align implicit selection with normal overload resolution.

I'd be happy to add some language to the spec and put together a write up for general consumption.

milessabin commented:
Nb. shapeless has been updated to compile once this is merged.

milessabin commented:
@lrytz rebased and issues addressed.

milessabin commented:
Rebased.

milessabin committed:
Make implicit selection consistent with normal overload resolution.

This is an attempt to make implicit selection rules align more closely
with people's intuitions and the spec's aspirations that it be
consistent with normal overload resolution rules. It also fixes
scala/bug#10526.

Candidate implicits are ordered via Infer#isStrictlyMoreSpecific which
delegates to Infer#isAsSpecific. Prior to this commit the latter treats all
methods with an implicit argument list as if they were nullary methods,
leading to scala/bug#10526 ... I believe that the motivation for this
behaviour was to allow methods which are intended to appear to users to
be nullary to have implicit arguments without affecting their overload
resolution behaviour wrt same-named methods which are actually nullary.
Whilst this does the right thing in some circumstances, it leads to
oddities such as scala/bug#10526 and also means that the specificity
test used to order and select implicits is very crude, in effect only
considering the result type of an implicit method, not its implicit
arguments.

This commit addresses these issues by making the "treat implicit methods
as nullary" logic explicitly flagged and enabling it only at the
relevant call site in Inferencer#inferExprAlternative. Normal
overloading rules are unchanged, and the implicit selection logic is
extracted out to ImplicitSearch#Improves. There the relative ordering of
a pair of candidate implicits is computed as follows,

1. The candidates are compared by their result type, the most specific by
   normal overload rules being selected (this is consistent with the
   behaviour prior to this commit).

2. If (1) is a tie then we compare the lengths of the implicit argument
   lists of the candidates (if any). The candidate with the longest
   argument list wins (this matches users intuitions that methods with more
   implicit arguments are "more demanding" hence more specific).

3. If (2) is a tie then we have a pair of candidates with implicit
   argument lists of equal length and which are both equally specific wrt
   their result type/the expected type. We now apply the full normal
   overloading resolution rules and choose the most specific.  This change
   makes combining implicits from multiple implicit scopes a great deal
   more straightforward, since we can now rely on more than just the result
   type of the implicit definition to guide implicit selection. With this
   commit the following works as expected,

  class Show[T](val i: Int)
  object Show {
    def apply[T](implicit st: Show[T]): Int = st.i

    implicit val showInt: Show[Int] = new Show[Int](0)
    implicit def fallback[T]: Show[T] = new Show[T](1)
  }

  class Generic
  object Generic {
    implicit val gen: Generic = new Generic
    implicit def showGen[T](implicit gen: Generic): Show[T] = new Show[T](2)
  }

  object Test extends App {
    assert(Show[Int] == 0)
    assert(Show[String] == 1)
    assert(Show[Generic] == 2) // showGen beats fallback due to longer argument list
  }

Thanks to the use of normal overload resolution rules we also have the
ability to use a "phantom" implicit parameter with arguments ordered by
subtyping to explicitly prioritize implicit definitions, something for
which language extensions of one form or another have occasionally been
proposed,

  class Low
  object Low {
    implicit val low: Low = new Low
  }
  class Medium extends Low
  object Medium {
    implicit val medium: Medium = new Medium
  }
  class High extends Medium
  object High {
    implicit val high: High = new High
  }

  class Foo[T](val i: Int)
  object Foo {
    def apply[T](implicit fooT: Foo[T]): Int = fooT.i

    implicit def foo[T](implicit priority: Low): Foo[T] = new Foo[T](0)
    implicit def foobar[T](implicit priority: Low): Foo[Bar[T]] = new Foo[Bar[T]](1)
    implicit def foobarbaz(implicit priority: Low): Foo[Bar[Baz]] = new Foo[Bar[Baz]](2)
  }
  class Bar[T]
  object Bar {
    implicit def foobar[T](implicit priority: Medium): Foo[Bar[T]] = new Foo[Bar[T]](3)
    implicit def foobarbaz(implicit priority: Medium): Foo[Bar[Baz]] = new Foo[Bar[Baz]](4)
  }
  class Baz
  object Baz {
    implicit def baz(implicit priority: High): Foo[Bar[Baz]] = new Foo[Bar[Baz]](5)
  }

  object Test extends App {
    assert(Foo[Int] == 0)
    assert(Foo[Bar[Int]] == 3)
    assert(Foo[Bar[Baz]] == 5)
  }

This commit changes the outcome of only one partest
(test/files/neg/t7289_status_quo.scala) which is intended to exactly
capture the current failure behaviour in a specific implicit resolution
scenario ... I believe the new (successful) behaviour is an improvement
on the old.

milessabin commented:
Rebased, fixed broken JUnit test, squashed, added useful commit message.

lrytz commented:
To me, this is still in need of a wider discussion - maybe @adriaanm can bring in a new perspective.

milessabin commented:
Arguably this is just aligning the implementation with the spec's aspirations, but whatever ... I'd be happy to write a short SIP up for this and have it discussed more widely. I guess I could also port to Dotty.

milessabin commented:
Rebased.

adriaanm commented:
We've discussed this in depth in our team meeting, and ran it by Martin. While we agree implicit prioritization can get tedious, we feel this PR introduces too much magic (e.g., longest argument list wins). The overall complexity of implicit resolution is already high enough. (The implicit argument list is skipped on purpose to avoid costly recursive searches.)

Ultimately, we are hopeful that we can replace type-level computation via implicit search prolog with a much more elegant, functional mechanism: [type-level meta-programming](https://github.com/dotty-staging/dotty/blob/10fee4d263de7c86bbcaf70f07a55fd2a8e14588/docs/docs/typelevel.md).

fommil commented:
@adriaanm I'd like to encourage you to reconsider closing this. The problem that is claimed to be fixed is huge and impacts a lot more codebases than I believe you to realise.

It feels unfair that the decision has been made behind closed doors. Don't we have a process for this so the community can decide how they want the language to evolve?

fommil commented:
@SethTisue https://contributors.scala-lang.org is for bike shedding. I'm not interested in wasting my time there.

I'm raising a Lightbend support ticket, as a commercial user of scala, to request that this be reconsidered. I don't see why I need to get involved any further as it's up to the stakeholders to decide and I am just one data point.

jvican commented:
@fommil https://contributors.scala-lang.org/ is precisely for this kind of discussions. It is not for bike shedding, and it's not only incorrect but disrespectful you say so for everyone investing their time in answering people's language proposals and creating proposals of their own.
 
This change needs to go through the SIP process in order to be merged: it's a language change of significant proportions. I don't see the point of raising a Lightbend support ticket. It's up to the original author of this proposal to create SIP, or to others to get involved in making that happen (or get the discussion started), to see further action.

adriaanm commented:
> It feels unfair that the decision has been made behind closed doors.

and

> I'm happy continuing the discussion via the private Lightbend portal

seem somewhat contradictory to me.

I'm also baffled that you keep complaining that you are somehow "forced into using a website that I don't enjoy using". We're perfectly happy to have a *technical* discussion on github, but please keep it on topic (complaining about discourse is not).

milessabin commented:
I'm not finding the non-technical commentary on this PR enlightening or helpful in any way. If noone has a good objection I'll delete it.

adriaanm commented:
You can also hide as off-topic, but either way is fine with me.

dwijnand commented:
:+1: for minimising as off-topic (and not deleting)

fommil commented:
@milessabin can we expect this branch to work as well as #6582 with `dist/mkPack` ? If so, I'd be interested in trying this out.

kailuowang commented:
Cats encounter this issue when migrating to 2.13. Without this PR it's yet to see how to fix without breaking our API. 
Detail in https://github.com/typelevel/cats/pull/2335#issuecomment-408236030
